### DISCUSSION TITLE
Workaround for Space toggling TreeView

### DISCUSSION THREAD

**Discussion by cristinathoughtpennies** [AvaloniaUI/Avalonia#9957] [Q&A]
Hello! I am at the latest, and I have a treeview of TextBoxes. I would like to be able to write text with spaces, but this causes the TreeViewItem to toggle expansion. Is there a way to turn off this shortcut for TreeViews?

**Comment by grokys** [comment]
This seems to have been caused by https://github.com/AvaloniaUI/Avalonia/pull/9598.

Maybe we shouldn't be using common keys like space and enter to toggle tree view items expansion. I notice the the WPF tree view has the same behavior but it's less noticeable because the tree view there only expands/collapses when using the numeric keypad afaics.

Ping @Glen-Nicol-Garmin @robloo

**Reply by timunie** [reply]
imo this should at least be configureable (opt-in or opt-out). In addition we may want to skip this shortcut if an input control has the keyboard focus. But this requires one additional check to be performed each time a key is pressed.

**Reply by robloo** [reply]
This is theoretically an issue with other controls depending on how you embed things.

Space/Enter are common actions and perhaps more useful on non-Windows platforms too. I also feel like TextBoxes in a TreeView are rare (usually read-only) so making it configurable may make sense too.

Ideally, the TextBox would just mark the event handled and if would be ignored by the tree view. Key events can tunnel though.

**Reply by glen-nicol** [reply]
I am open to making changes. But I am not aware of a standard way to make keybindings dynamic in avalonia. I thought I saw someone mention it as a future feature to come up with a system for that. Sadly, my time at work has shifted to another higher priority so I won't be working with avalonia every day for a few months. So I won't be able to help brainstorm that level of work.

Sounds like you are doing something like a file tree and want to allow renames(like explorer or visual studio solution explorer)? Or do you need to allow typing at all times? If its a rename mode then could you use a popup that overlays the treeviewitem header or something similar to escape the treeview temporarily? Not sure how VS does it.  To be honest I haven't had to do enough with bubble and tunnel to have a firm grasp. But yeah seems like maybe our events are just not quite as tuned/configured for this use case.

I think I would be in favor of removing the space handler over adding a property to turn it on or off. But enter seems pretty foundational (and is used in WPF) so don't want to remove that one. So maybe we need a property that takes the key enum as a csv to mask in enabled/disabled keys?

**Reply by cristinathoughtpennies** [reply]
It's a dialogue tree, so I need to allow typing at all times

**Reply by robloo** [reply]
Note that a lot of the WinUI controls (like ComboBox, Button, etc.) support space and enter to do these sorts of actions. The intention when I brought that up in Glen's PR was to modernize a few things here and make this functionality more obvious than number pad multiply/divide.

It is a general problem though. If you put a TextBox in a SplitButton, ToggleSplitButton or even a regular old Button in WinUI you run across this same issue. Avalonia inherits this problem.

I think it should be considered for all cases before just removing space support in tree view. A more general solution is needed.

**Reply by cristinathoughtpennies** [reply]
Is there something I can do in the meanwhile to work around this?

**Reply by amwx** [reply]
I think if anything, Space (and any other key that inputs text) needs to be handled with the TextInput event rather than KeyDown. Event sequence for `Space` is:

tunnel KeyDown
OnKeyDown                   <- this is where TreeView is handling the event now
bubble KeyDown
tunnel TextInputEvent     
OnTextInput                    <- this is where TextBox will try handle it
bubble TextInputEvent

Handling in TextInput (probably bubbled event path, not sure if the class handler OnTextInput is enough) should allow seeing if there's a control taking text input which marks the event handled, and then the TreeView (or any other control) can decide whether it should still handle the event. The downside is this breaks up key handling into two different places.

The other solution would be to refactor how KeyDown<->TextInput works. Interestingly, UWP dropped this event and I'm assuming they just wrap it into the key handler, but as robloo pointed out the issue still exists. I also have no idea how this would be done and what the side effects would be.

**Reply by glen-nicol** [reply]
@cristinathoughtpennies for now I think you can subclass TreeViewItem and override OnKeyDown. Then you could check to see if the pressed key is space and if it is then don't call the base method and instead mark it handled. That will solve the key handling. But if you're using databinding then you will likely also need to change the hosting treeview ItemContainerGenerator to create objects of your subclass instead of the normal treeviewitems. I do not actually know how to do that.

**Reply by cristinathoughtpennies** [reply]
@glen-nicol Ok, I'll look into doing that, thanks for the suggestion

**Reply by cristinathoughtpennies** [reply]
In the meanwhile, should a bug be made for this?

**Reply by cristinathoughtpennies** [reply]
So, I want to override the double click for a similar reason, but because HeaderDoubleTapped is private, I can't.
@glen-nicol would you be willing to change the function to be protected instead?

**Reply by robloo** [reply]
@cristinathoughtpennies It's usually better to file an issue and/or start a PR for such changes yourself. You will usually get feedback and guidance on any PR even if you are unsure of something.

It certainly might make sense to have a `protected virtual void OnHeaderDoubleTapped(TappedEventArgs e)` method.

**Reply by glen-nicol** [reply]
I agree that making that function protected and virtual is a good idea. Would you be able to do the PR? Might be faster than my turnaround time and I think it will be pretty straightforward. I think you just have to change the modifiers and maybe move the function up so it's with the other protected functions.

**Reply by cristinathoughtpennies** [reply]
Gotcha, here is the PR! https://github.com/AvaloniaUI/Avalonia/pull/10022
